oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/oidc/authentication/OidcAuthorizedClientRefreshedEventListener.java

```diff
@@ -0,0 +1,219 @@
+ * {@link OAuth2AuthorizedClientRefreshedEvent} and publishes an event of type
+ * {@link OidcUserRefreshedEvent} in order to refresh an {@link OidcUser}.
+ *
+ * @author Steve Riesenberg
+ * @since 6.5
+ * @see org.springframework.security.oauth2.client.RefreshTokenOAuth2AuthorizedClientProvider
+ * @see OAuth2AuthorizedClientRefreshedEvent
+ * @see OidcUserRefreshedEvent
+ */
+public final class OidcAuthorizedClientRefreshedEventListener
+		implements ApplicationEventPublisherAware, ApplicationListener<OAuth2AuthorizedClientRefreshedEvent> {
+
+	private static final String INVALID_ID_TOKEN_ERROR_CODE = "invalid_id_token";
+
+	private static final String INVALID_NONCE_ERROR_CODE = "invalid_nonce";
+
+	private OAuth2UserService<OidcUserRequest, OidcUser> userService = new OidcUserService();
+
+	private JwtDecoderFactory<ClientRegistration> jwtDecoderFactory = new OidcIdTokenDecoderFactory();
+
+	private GrantedAuthoritiesMapper authoritiesMapper = (authorities) -> authorities;
+
+	private SecurityContextHolderStrategy securityContextHolderStrategy = SecurityContextHolder
+		.getContextHolderStrategy();
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	@Override
+	public void onApplicationEvent(OAuth2AuthorizedClientRefreshedEvent event) {
+		if (this.applicationEventPublisher == null) {
+			return;
+		}
+
+		// The response must contain the openid scope
+		OAuth2AccessTokenResponse accessTokenResponse = event.getAccessTokenResponse();
+		if (!accessTokenResponse.getAccessToken().getScopes().contains(OidcScopes.OPENID)) {
+			return;
+		}
+
+		// The response must contain an id_token
+		Map<String, Object> additionalParameters = accessTokenResponse.getAdditionalParameters();
+		if (!StringUtils.hasText((String) additionalParameters.get(OidcParameterNames.ID_TOKEN))) {
+			return;
+		}
+
+		// The current authentication must be an OAuth2AuthenticationToken
+		Authentication authentication = this.securityContextHolderStrategy.getContext().getAuthentication();
+		if (!(authentication instanceof OAuth2AuthenticationToken authenticationToken)
+				|| authenticationToken.getClass() != OAuth2AuthenticationToken.class) {
+			// This event listener only handles the default authentication result. If the
+			// application customizes the authentication result, then a custom event
+			// handler should be provided.
+			return;
+		}
+
+		// The current principal must be an OidcUser
+		if (!(authenticationToken.getPrincipal() instanceof OidcUser existingOidcUser)) {
+			return;
+		}
+
+		// The registrationId must match the one used to log in
+		ClientRegistration clientRegistration = event.getAuthorizedClient().getClientRegistration();
+		if (!authenticationToken.getAuthorizedClientRegistrationId().equals(clientRegistration.getRegistrationId())) {
+			return;
+		}
+
+		// Refresh the OidcUser and send a user refreshed event
+		OidcIdToken idToken = createOidcToken(clientRegistration, accessTokenResponse);
```

Comment:
Shouldn't we do this only if the current authentication matches the principal in the authorized client in the event? Not sure if it can happen that a refresh happens the client registration is the same, but the current authentication does not match the authorized client principal

Reply:
Thanks for the feedback @filiphr.

Actually, at the point where this event is being processed, the principal name in the authorized client will always be the same as the current authentication. The reason for this is that the authorized client is typically created from the current principal using the same `SecurityContextHolderStrategy` to resolve it. This would only not be the case if the application is customized to somehow populate the principal differently, which I think would be quite uncommon. In any case, if we added a check for it here, it would always pass and so I'm not sure it makes much sense.

Reply:
You are right @sjohnr. It indeed does not make sense that when using the same registration as the one that the login one the principal in the authorized client is different as the one from the authentication.

I think that I missed the prior check and got side tracked with my thoughts.